android/java/org/chromium/chrome/browser/search_engines/settings/BraveBaseSearchEngineAdapter.java

```diff
@@ -82,4 +85,68 @@ public static void sortAndFilterUnnecessaryTemplateUrl(
             return SearchEngineAdapter.TemplateUrlSourceType.RECENT;
         }
     }
+
+    // Fields and empty methods below are to satisfy java compiler and will be
+    // removed by bytecode asm
+    public List<TemplateUrl> mPrepopulatedSearchEngines = new ArrayList<>();
+    public List<TemplateUrl> mRecentSearchEngines = new ArrayList<>();
+    public int mSelectedSearchEnginePosition = -1;
+
+    public static boolean containsTemplateUrl(
+            List<TemplateUrl> templateUrls, TemplateUrl targetTemplateUrl) {
+        assert false;
+        return true;
+    }
+
+    public int computeStartIndexForRecentSearchEngines() {
+        assert false;
+        return -1;
+    }
+
+    public boolean didSearchEnginesChange(List<TemplateUrl> templateUrls) {
+        if (templateUrls.size()
+                != mPrepopulatedSearchEngines.size() + mRecentSearchEngines.size()) {
+            return true;
+        }
+        for (int i = 0; i < templateUrls.size(); i++) {
+            TemplateUrl templateUrl = templateUrls.get(i);
+            if (!containsTemplateUrl(mPrepopulatedSearchEngines, templateUrl)
+                    && !containsTemplateUrl(mRecentSearchEngines, templateUrl)) {
+                return true;
+            }
+        }
+        // Till this line it is a straight copy of SearchEngineAdapter.didSearchEnginesChange.
```

Comment:
We should call `SearchEngineAdapter.didSearchEnginesChange` via `BraveReflectionUtil` instead of copying code, see `createViewProvider` as an example.

Reply:
Thanks, there was a problem in `BraveBaseSearchEngineAdapter.didSearchEnginesChange`. In fact the method always gave true, because it's fields `mPrepopulatedSearchEngines/mRecentSearchEngines/mSelectedSearchEnginePosition` were duplicated in both `BraveBaseSearchEngineAdapter` and `SearchEngineAdapter`. Fixed by using `BraveReflectionUtil`.

Fixed at https://github.com/brave/brave-core/pull/14402/commits/22d3cd2546b3d11d11a1785792e4d6482e83dee8